### PR TITLE
RequestDelegateFactory context classes ApplicationServices property

### LINKED ISSUE #41900: Add access to services for route handler filter factories to RouteHandlerContext
When adding a route handler filter via a filter factory there are cases where the filter factory (as opposed to the filter itself) needs access to the application services, e.g. to get an `ILogger` instance. We should add a property to `RouteHandlerContext` to enable this.

```diff
namespace Microsoft.AspNetCore.Http;

public sealed class RouteHandlerContext
{
+    public IServiceProvider ApplicationServices { get; }
}
```

Note that the property is nullable due to where it's instantiated in `RequestDelegateFactory` and matches the `Services` properties on `EndpointMetadataContext` and `EndpointParameterMetadataContext`.

## Example Usage

```csharp
public static RouteHandlerBuilder WithParameterValidation(this RouteHandlerBuilder builder, int statusCode = StatusCodes.Status400BadRequest)
{
    builder.Add(eb =>
    {
        var methodInfo = eb.Metadata.OfType<MethodInfo>().SingleOrDefault();
        if (methodInfo is not null && IsValidatable(methodInfo))
        {
            eb.Metadata.Add(new ProducesResponseTypeAttribute(typeof(HttpValidationProblemDetails), statusCode, "application/problem+json"));
        }
    });
    builder.AddFilter((RouteHandlerContext context, RouteHandlerFilterDelegate next) =>
    {
        // *************************************
        // ** We use ApplicationServices here **
        // *************************************
        var loggerFactory = context.ApplicationServices.GetRequiredService<ILoggerFactory>();
        var logger = loggerFactory.Create("MinimalApis.Extensions.Filters.ValidationRouteHandlerFilterFactory");
 
        if (!IsValidatable(context.MethodInfo))
        {
            return next;
        }
 
        return rhic =>
        {
            foreach (var parameter in rhic.Parameters)
            {
                if (parameter is not null && !MiniValidator.TryValidate(parameter, out var errors))
                {
                    return new(Results.ValidationProblem(errors));
                }
            }
 
            return next(rhic);
        };
    });

    return builder;
}
```

## Risks

Should be very low. This is following the pattern used in other context classes including those where the call site originates in `RequestDelegateFactory`, e.g. `EndpointMetadataContext`.

### PR DESCRIPTION
**Pull request by DamianEdwards**
- Added `public IServiceProvider? Services { get; }` property to `RouteHandlerContext`
- Enables route handler filter factories to access app services
- Added a unit test & updated another to ensure app's service provider is passed through correctly by `RequestDelegateFactory`
- Renamed `EndpointMetadataContext.Services` and `EndpointParameterMetadataContext.Services` to `ApplicationServices` and made non-nullable
- Added unit tests to ensure context `ApplicationServices` property is set correctly

Fixes #41900

### CODE REVIEW

File: src/Http/Http.Abstractions/src/RouteHandlerContext.cs
Comment (davidfowl) [inline src/Http/Http.Abstractions/src/RouteHandlerContext.cs:20]: Why is it nullable?

Comment (DamianEdwards) [reply src/Http/Http.Abstractions/src/RouteHandlerContext.cs:20]: Because it's nullable on `RequestDelegateFactoryOptions` which is where it's set from. Same thing we did for `EndpointMetadataContext` and `EndpointParameterMetadataContext`.

File: src/Http/Http.Abstractions/src/RouteHandlerContext.cs
Comment (halter73) [inline src/Http/Http.Abstractions/src/RouteHandlerContext.cs:24]: Nit:

```suggestion
        ArgumentNullException.ThrowIfNull(methodInfo);
        ArgumentNullException.ThrowIfNull(endpointMetadata);
        ArgumentNullException.ThrowIfNull(applicationServices);
```

The second parameter uses `[CallerArgumentExpression("argument")]`.

File: src/Http/Routing/test/UnitTests/Builder/RouteHandlerEndpointRouteBuilderExtensionsTest.cs
Comment (halter73) [inline src/Http/Routing/test/UnitTests/Builder/RouteHandlerEndpointRouteBuilderExtensionsTest.cs:996]: Nit:

```suggestion
        Assert.Single(dataSource.Endpoints);
        Assert.True(filterRan);
```

It'd be nice to have a bool that gets set in the callback so we can do the assert above.

**Review by halter73 — APPROVED** [review]
We should wait for API approval just in case it changes of course.

**Review by captainsafia — APPROVED** [review]

**Pull request merged**
